Use parent for bfs validation

nicelhc13 opened the pull request:
Since other variants than direction optimization are using distance labels, CI test would be failed.

scober commented:
It turns out '_' was ignored on purpose.

ddn0 commented:
FWIW, `_` => "" would be fine too, you just need to run `scripts/check_ifndef.py -fix lib*` as well.

scober commented:
> FWIW, `_` => "" would be fine too, you just need to run `scripts/check_ifndef.py -fix lib*` as well.

As the original author, do you have a preference?

tylershunt commented:
If we were to handle `"_"` in a special way I would rather make it harder to shadow files, for instance turning it into `"__"` (double underscore) rather than striping away the difference between camel case and snake case.

scober commented:
I like that.

ddn0 commented:
The original reason I didn't handle `_` (specially) was because I didn't want collisions between guards of different files. Keeping the mapping one-to-one guarantees that.

Escaping is the correct answer, though it looks kind of ugly in practice. (At some point, baseXY encoding the path becomes a viable answer too.)

- `_` and `-` => "" is reasonable because it is simple to explain and remember
- `-` => "", `_` => `_`  also seems simple enough
- `_` => `__` seems more borderline to me. At this point, I begin to think `#pragma once` is also okay. Since this is automated, the pain of changing policies is low, so I don't have a strong opposition to this choice either.

ddn0 commented:
Also, technically
- the classes `LC_CSR_XYZ` are going to be refactored with new property graph support anyway, and
- `Executor_ForEach.h` and co could equally be called `ExecutorForEach.h`
So `_` can be less common in names.

ddn0 commented on scripts/check_ifndefs.py:
```diff
@@ -49,6 +49,10 @@ def make_guard(root, filename):
+    # Just in case, remove characters that can't be part of macros
+    p = re.sub("[+\-*%=<>?~&\^|#:;{}.[\]]","", p)
+    # Differentiate between snake_case file names and directories
+    p = p.replace("_", "__", 1)
```
`p.replace(..., -1)` to replace all occurrences?

tylershunt commented:
> The original reason I didn't handle `_` (specially) was because I didn't want collisions between guards of different files. Keeping the mapping one-to-one guarantees that.

That makes sense as a goal but I would stop short of saying that the identity mapping prevents collisions. For instance the current file `libtsuba/include/tsuba/s3_internal.h` would collide with `libtsuba/include/tsuba/s3/internal.h`. That seems ok since we don't create deep nesting in the headers. What I was more concerned about were cases like `libtsuba/src/tsuba_internal.h` which (I think) collides with `libtsuba/include/tsuba/internal.h`

Yeah I take your point about `pragma once`

scober commented:
Is the concern about `#pragma once` that we want to support building in environments that don't offer it?

scober commented:
I think I like `_` => "" because "don't have two files in the same directory with the same name with one in camel case and one in snake case" seems like a reasonable policy and would eliminate the only(?) collision.

ddn0 commented:
I think `#pragma once` is just us old grognards wondering why we need all these new fangled things when back in our day we used m4 and liked it. And technically it is not standardized and future C++ module support will probably supersede it, so inertia might be beneficial here.

I don't think we have a concrete platforms that we intend to support that don't also support `#pragma once` too.

Commit:
Use parent for bfs validation

In the past, different nod labels were used for different bfs varaints.
For example, only direction-optimization algorithm used a parent
id for its node label, and did not use/calculate distance from
a start node. But, all variants passed through the same validation
phase. This commit integrates node labels to its parent to be validated.

arthurp commented on scripts/bench_python_cpp_algos.py:
```diff
@@ -36,8 +36,8 @@ def run_bfs(property_graph: PropertyGraph, input_args, source_node_file):
-    if "road" in input_args["name"]:
-        bfs_plan = analytics.BfsPlan.asynchronous()
+    # if "road" in input_args["name"]:
+    #        bfs_plan = analytics.BfsPlan.asynchronous()
```
Remove instead of comment.

arthurp reviewed (commented):
This will be rebased on top of your other parent PR?

nicelhc13 commented:
> This will be rebased on top of your other parent PR?

No, I think that PR is still under discussion and is not clear to me yet. As @roshandathathri  suggested, our focus for now is to make direction-optimization works.

arthurp commented:
> > This will be rebased on top of your other parent PR?
> 
> No, I think that PR is still under discussion and is not clear to me yet. As @roshandathathri suggested, our focus for now is to make direction-optimization works.

This PR doesn't really make any sense without the other changes. So this will also need to wait until we figure out the plan for BFS and SSSP outputs.

roshandathathri commented on libgalois/src/analytics/bfs/bfs.cpp:
```diff
@@ -500,33 +500,82 @@ katana::analytics::Bfs(
-    return katana::ErrorCode::AssertionFailed;
+  for (uint32_t u : graph) {
+    // TODO(lhc): should be BfsParent
+    //            (keep this since still exist variants using distance as label)
```
Change "since still exist" to "since there exists"

roshandathathri commented on libgalois/src/analytics/bfs/bfs.cpp:
```diff
@@ -500,33 +500,82 @@ katana::analytics::Bfs(
+  for (uint32_t u : graph) {
+    // TODO(lhc): should be BfsParent
+    //            (keep this since still exist variants using distance as label)
+    uint32_t u_parent = graph.GetData<BfsNodeDistance>(u);
```
Why is "BfsNodeDistance" a parent? Rename that to "BfsNodeParent".

nicelhc13 replied:
This was b/c we didn't decide parent vs distance. Other variants have used distance. Now it is good to change to parent.

roshandathathri reviewed (approved):
LGTM but let's wait for @arthurp's approval (and changes)

arthurp reviewed (changes requested):
* Please disable the other algorithms in this PR as well. You can just comment out the constructors on the plan (in C++ and Python) and change the default to the algorithm we are leaving.
* Update the documentation to match what we are really computing.

arthurp commented on libgalois/include/katana/analytics/bfs/bfs.h:
```diff
@@ -95,9 +95,7 @@ struct KATANA_EXPORT BfsStatistics {
-  uint32_t max_distance;
-  /// The average distances on visited nodes from the source node.
-  double average_visited_distance;
+  uint32_t max_parent;
```
What is the `max_parent` and why would I care? Maybe just drop this entirely and we can put meaningful stats back in a bit later. 

Also docs need updating here.

nicelhc13 replied:
Hehe yeah, let me remove this! Yeah I am gonna update documents.

roshandathathri replied:
Agree. Not useful.

arthurp reviewed (changes requested):
What happens if I call with an unsupported plan? For now we could just return `katana::ErrorCode::NotImplemented`. Currently it looks like it would run the algorithm which will do the wrong thing.

arthurp commented on python/katana/analytics/_bfs.pyx:
```diff
@@ -190,7 +190,7 @@ def bfs_assert_valid(PropertyGraph pg, str property_name):
     output_property_name_cstr = <string>output_property_name_bytes
     with nogil:
-        handle_result_assert(BfsAssertValid(pg.underlying_property_graph(), output_property_name_cstr))
+        handle_result_assert(BfsAssertValid(pg.underlying_property_graph(), start_node, output_property_name_cstr))
```
start_node isn't an argument to this function at the Cython level. You will need to add it above. And add it in the test cases.

arthurp replied:
I'm fixing this. Commit in coming.

arthurp reviewed (approved):
My issues are addressed. I'm gonna be AFK for a while. So I'm gonna approve to avoid blocking anything.

roshandathathri reviewed (approved):
LGTM too

ddn0 committed:
Block BC sources per trial in benchmark script

roshandathathri reviewed (approved):
LGTM
